### DISCUSSION TITLE
Modeling Maps

### DISCUSSION THREAD

**Discussion by ericzbeard** [apple/pkl#311] [Q&A] [answered]
I'm wondering if it's possible to have a strongly typed map that lets me use object syntax, so I can avoid the extra typing involved with `["Key"]` and just type `Key`, but the type of `Value` is still enforced. (To be clear, I need a map because the names of the keys are not known ahead of time)

Here's an example:

```pkl
class Widget {
    Name: String
}

class A {
    Props: Mapping<String, Widget>
}

// Valid
a: A = new {
    Props {
        ["Foo"] {
            Name = "widget a"
        }
    }
}

// Valid
aa: A = new {
    Props {
        ["Foo"] = new Widget {
            Name = "widget aa"
        }
    }
}

// Invalid but this is what I want (less verbose)
aaa: A = new {
    Props {
        Foo {
            Name = "widget aaa"
        }
    }
}
```

**Comment by bioball** [comment, marked as answer]
This isn't possible; it was actually an intentional design goal to separate the grammar for these two. The intention is that you can unambiguously tell which keys are not fixed when reading through Pkl code.

**Reply by ericzbeard** [reply]
Thanks! Makes sense, just making sure I didn't miss anything.

**Reply by holzensp** [reply]
That said, you *can* fake it. If all of your values are of type `Foo`, you can say `lightSyntaxFakeMapping: Dynamic(toMap().values.every((it) -> it is Foo))`

**Reply by bioball** [reply]
That’s possible, although, I wouldn’t recommend it. 

With this pattern, it’s not possible to define `default`, because it already exists on the class. 

You also don’t have niceties like autocompletion and checking in the IDE.
